### ISSUE TITLE
newsboat uses only a single core

### ISSUE THREAD

**Issue by emanruse** [newsboat/newsboat#2815]
**Newsboat version (copy the output of `newsboat -v` or the first line of `git show`)**:

```
$ git show
commit 40828e6e8541d3a4f6b9a9faca88000ba9c96741 (HEAD -> master, origin/master, origin/HEAD)
```

and as it seems all earlier versions.

Config file (copy from ~/.newsboat/config or ~/.config/newsboat/config):

```
# irrelevant
```

Steps to reproduce the issue:

- Refresh feeds

or

- Populate a query feed (e.g. with a lot of feeds and complex expressions)

or 

- something other taking a fairly long time, e.g. vacuum.

While doing this, watch system load (e.g. `load average`)

### Result

newsboat uses only a single core.

**Other info you think is relevant**:

Related to #2350

**Comment by Minoru** [comment]
Refreshing feeds can be made faster with [`reload-threads`](https://newsboat.org/releases/2.36/docs/newsboat.html#reload-threads).

Query feeds are covered by #2350 that you've linked to.

Vacuum is performed by SQLite itself and to my knowledge it can't be sped up.

What other operations are slow? I can think of search, where we could maybe make it faster by building an index ourselves rather than rely on SQLite's `LIKE` operator; or maybe look into full-text search extensions.

**Comment by Minoru** [comment]
> Last time I tested (months or a year ago), it didn't have any effect. Tested with local xml feeds, i.e. not due to network bottleneck. Could you suggest a different testing procedure?

I can't reproduce. My config and urls file: [configs.tar.gz](https://github.com/user-attachments/files/16553007/configs.tar.gz). When reloading with `newsboat -x reload`, I observe 8 Newsbaot threads in `htop` which disappear after about 6 seconds, and the program finishes.

> This has suddenly become extremely slow

Can confirm: for my real config and urls file, the reload time went from 1m40s to 2m30s. Bisect points to 8eade078760c769715b19bff7f41aa2b83efd19d, which doesn't make sense to me because that crate is only used in tests =\ I should re-test in a better-controlled environment, because I'd expect https://github.com/newsboat/newsboat/pull/2804 to cause a slow-down, but not *this large* of a slowdown.

**Comment by dennisschagt** [comment]
> > This has suddenly become extremely slow
> 
> Can confirm: for my real config and urls file, the reload time went from 1m40s to 2m30s. Bisect points to [8eade07](https://github.com/newsboat/newsboat/commit/8eade078760c769715b19bff7f41aa2b83efd19d), which doesn't make sense to me because that crate is only used in tests =\ I should re-test in a better-controlled environment, because I'd expect #2804 to cause a slow-down, but not _this large_ of a slowdown.

Depending on how you tested, there might have been a big variance caused by cached feeds due to last-modified/etag handling (which would skip conversion to utf-8 and xml parsing).
Next to that, if you were testing quickly, you might have run into rate limiting on some feeds (possibly getting HTTP 429 Too Many Requests, which would also result in skipped conversions and parsing).

Anyway, I believe I've found the issue.
We recently started using `convert_text()`.
Apparenty, that uses a small buffer (16 bytes), causing a high number of iterations of calling `iconv()`.
Increasing the buffer size seems to make it about 75 times faster (in the case of a 520 KB file, using an output buffer of 64KB instead of 16 B).
See https://github.com/newsboat/newsboat/pull/2822 for more info.

**Comment by Minoru** [comment]
Yeah, that's why I wanted to redo the tests in a better-controlled environment ;)

As for the original issue with slow reload, I have yet another idea. Assuming we're talking about `newsboat -x reload` and not the `reload` operation in TUI, it might be that the slowness comes from the database rather than reload itself. Just now I ran `newsboat -x reload` for the first time since starting the computer, and here's what I observed in `htop`:
1. for about a minute `newsboat` was using 70-90% of one core (my cache.db is 5.5G);
2. then for a brief moment there were as much Newsboat threads as I set in my config;
3. then for a few more minutes one thread was pegged at 100% while a handful others were in 5-40% range.

What I think is going on here:
1. Newsboat is reading the database, applying ignore rules as it goes;
2. Newsboat is reloading the feeds (d'oh!);
3. Newsboat is saving feeds to the database.

The access to the database is sequential, i.e. only one thread can read or write in any given moment. For the first startup, it is also somewhat limited by I/O (#766), but not entirely. The linked issue has the best ideas on how to improve that.

Also, SQLite has a so-called ["WAL mode"](https://www.sqlite.org/wal.html) which lets multiple readers run concurrently. This could help with (1) in a sense that multiple threads could read different feeds simultaneously, saturating the I/O. It could also help with (3) if we did `Cache::externalize_rssfeed()` in two steps:
1. given a feed to save, pick the items that either don't exist in the DB or differ in some way from the ones in the DB;
2. upgrade to write mode and save just the picked items.

I don't know how much time this could actually save, but it minimizes the time that a thread has to hold the (exclusive) write lock. Also, it might make sense to do this even now (with a single thread and in journal mode), because writes are amplified (data is written to the journal *and* the database), while reads don't have that penalty; so it might be plain faster to read data and skip the write rather than blindly overwriting the data.
